packages/alfa-rules/test/sia-er62/rule.spec.tsx

```diff
@@ -1070,229 +1225,255 @@ test(`evaluate() passes an <a> element that is distinguishable from the <p> pare
+/******************************************************************
+ *
+ * Inapplicable tests
+ *
+ ******************************************************************/
+test(`evaluate() is inapplicable to an <a> element with no visible text content`, async (t) => {
+  const target = (
+    <a href="#">
+      <span hidden>Link</span>
+    </a>
+  );
+
+  const document = h.document([<p>Hello {target}</p>]);
+
+  t.deepEqual(await evaluate(ER62, { document }), [inapplicable(ER62)]);
+});
+
+test(`evaluate() is inapplicable to an <a> element with a <p> parent element
+      no non-link text content`, async (t) => {
+  const target = <a href="#">Link</a>;
+
+  const document = h.document([<p>{target}</p>]);
+
+  t.deepEqual(await evaluate(ER62, { document }), [inapplicable(ER62)]);
+});
+
+test(`evaluate() is inapplicable to an <a> element with a <p> parent element
+      no visible non-link text content`, async (t) => {
+  const target = <a href="#">Link</a>;
+
+  const document = h.document([
+    <p>
+      <span hidden>Hello</span> {target}
+    </p>,
+  ]);
+
+  t.deepEqual(await evaluate(ER62, { document }), [inapplicable(ER62)]);
+});
+
+test(`evaluate() is inapplicable to an <a> element with a <p> parent element
+      whose role has been changed`, async (t) => {
+  const target = <a href="#">Link</a>;
+
+  const document = h.document([
+    <p role="generic">
+      <span>Hello</span> {target}
+    </p>,
+  ]);
+
+  t.deepEqual(await evaluate(ER62, { document }), [inapplicable(ER62)]);
+});
+
+test(`evaluate() is inapplicable to an <a> element with a <p> parent element
+    no non-link whitespace text content`, async (t) => {
+  const target = <a href="#">Link</a>;
+
+  const document = h.document([<p> {target}</p>]);
+
+  t.deepEqual(await evaluate(ER62, { document }), [inapplicable(ER62)]);
+});
+
+test(`evaluate() is inapplicable to an <a> element with a <p> parent element
+    when both have the same foreground color`, async (t) => {
   const target = <a href="#">Link</a>;
 
   const document = h.document(
     [<p>Hello {target}</p>],
     [
       h.sheet([
         h.rule.style("p", {
-          backgroundImage: "linear-gradient(to right, #F9F9F1 50%, blue 50%)",
-          color: "rgba(255, 255, 255, 0.1)",
+          color: "#0000EE",
         }),
+      ]),
+    ]
+  );
 
-        h.rule.style("a, a:hover, a:focus", {
-          backgroundImage: "linear-gradient(to right, #F9F9F1 50%, blue 50%)",
+  t.deepEqual(await evaluate(ER62, { document }), [inapplicable(ER62)]);
+});
+
+test(`evaluate() is inapplicable to an <a> element with a <p> parent element
+    when both have the same foreground color and <a> element's sibling elements also have the same foreground color`, async (t) => {
+  const target = <a href="#">Link</a>;
+
+  const document = h.document(
+    [
+      <div>
+        <p>Hello</p>
+        <p>
+          <span>World</span>
+          Hello
+          <p>nested p</p>
```

Comment:
Nested `<p>` are not allowed in the DOM. Not sure whether we enforce that during parsing (I guess not), but we should never have that as input (UA should fix the code and build a correct DOM, likely moving the nested `<p>` around).

(nested `<div>` or `<span>` work fine to test deep trees, of course)

Reply:
Because we need to draw the line somewhere 😅 and we need something to compare to.

---
[WCAG Success Criterion 1.4.1 Use of Color](https://www.w3.org/TR/WCAG21/#use-of-color) states:
> Color is not used as the only visual means of conveying information, indicating an action, prompting a response, or distinguishing a visual element.

So, for the rule to correctly map to it (i.e. to avoid false positives where the rule would fail but WCAG would be satisfied), we need to be (reasonably) "sure" that color is the **only** visual mean to convey the information "this is a link"…

There are cases, such as breadcrumbs or a site map where that info is conveyed just by being text in the breadcrumb, or list item in the site map. So we can't look at all links and flag all of them.

While I do agree that a `<div>`  works more or less like a `<p>` and we could flag these, we need to draw the line on the safe side. By restricting the rule to link inside paragraphs, we can be fairly sure that there is nothing more than style to convey the information "this is a link" (this excludes icons such as 🔗 that are sometimes used to prefix link, there is an assumption about that in the rule, we're not seeing too many complains about them, so the assumption seems to mostly hold).

So, we work in a frame where we can be confident we report true problems.

---
And of course, if there is no surrounding text at all, there is nothing to compare to 😃 

We could totally look at links inside `<div>`, but `<div>` may be used for more or less anything. Role of `paragraph` has the nice semantic linking, that these should only contain "block of text" where the rule make perfect sense.

---
The "no false positives" motto is putting strong constraints and often lead us to write weaker rules 😞

Reply:
Thank you, this explains it all :)